### PR TITLE
[Client-1176] Added support for write operations in background queries

### PR DESCRIPTION
**Pull request by vmsachin**

### CODE REVIEW

File: lib/aerospike/policy/write_policy.rb [hunk 1]
```diff
@@ -74,6 +74,20 @@ def initialize(opt={})
       # Valid for Aerospike Server Enterprise Edition 3.10+ only.
       @durable_delete = opt.fetch(:durable_delete, false)
 
+      @records_per_second = opt[:records_per_second] || 0
```
Comment (khaf) [inline lib/aerospike/policy/write_policy.rb:77]: No documentation here?

File: lib/aerospike/client.rb [hunk 2]
```diff
@@ -731,6 +730,59 @@ def query(statement, options = nil)
...
+    # If the policy is nil, the default relevant policy will be used.
+    #
+    # @param statement [Aerospike::Statement] The query or batch read statement.
+    # @param operations [Array<Aerospike::Operation>] An optional list of operations.
+    # @param options [Hash] An optional hash of policy options.
+    # @return [Aerospike::ExecuteTask] An ExecuteTask instance that can be used to wait for command completion.
+    #
+    # @raise [Aerospike::Exceptions::Aerospike] if an error occurs during the operation.
+    def query_execute(statement, operations = [], options = nil)
+      policy = create_policy(options, WritePolicy, default_write_policy)
```
Comment (khaf) [inline lib/aerospike/client.rb:753]: I'd rather see two policies here: One for WritePolicy and one for QueryPolicy. That way you'll not have to put new field in WritePolicy that are only related and applicable to the query_execute command.

File: lib/aerospike/policy/write_policy.rb [hunk 3]
```diff
@@ -74,6 +74,20 @@ def initialize(opt={})
...
+
+      # Determines network timeout for each attempt.
+      #
+      # If socket_timeout is not zero and socket_timeout is reached before an attempt completes,
+      # the Timeout above is checked. If Timeout is not exceeded, the transaction
+      # is retried. If both socket_timeout and Timeout are non-zero, socket_timeout must be less
+      # than or equal to Timeout, otherwise Timeout will also be used for socket_timeout.
+      #
+      # Default: 30s
+      @socket_timeout = opt[:socket_timeout] || 30000
```
Comment (khaf) [inline lib/aerospike/policy/write_policy.rb:87]: This should inherit from the base policy. All commands have socket_timeout.

File: spec/aerospike/query_spec.rb [hunk 4]
```diff
@@ -87,7 +87,7 @@
         count = 0
         rs.each do |rec|
           count += 1
-          expect(rec.bins.keys).to contain_exactly("bin1", "bin2", "bin3", "bin4")
+          expect(rec.bins.keys).to contain_exactly("bin1", "bin2", "bin3", "bin4", "bin5")
```
Comment (khaf) [inline spec/aerospike/query_spec.rb:90]: Why is this test affected by a test coming after?

File: spec/aerospike/query_spec.rb [hunk 5]
```diff
@@ -512,6 +512,20 @@
...
     end # context
 
+    context "Run queries in the background" do
+      it "executes a put operation asynchronously" do
+        stmt = Aerospike::Statement.new(@namespace, @set)
+        bin5 = Aerospike::Bin.new("bin5", "string_value")
+        ops = [
+          Aerospike::Operation.put(bin5)
+        ]
+        async_task = client.query_execute(stmt, ops)
```
Comment (khaf) [inline spec/aerospike/query_spec.rb:522]: I'd rather this be called execute_task. No reason to convey the client has an async capability that it doesn't.

**Review by khaf — CHANGES_REQUESTED** [review]

File: lib/aerospike/client.rb (hunk 2 above)
Comment (vmsachin) [reply lib/aerospike/client.rb:753]: moved records_per_second to statement similar to the java client.

File: spec/aerospike/query_spec.rb (hunk 4 above)
Comment (vmsachin) [reply spec/aerospike/query_spec.rb]: Have reverted the changes!

File: lib/aerospike/policy/write_policy.rb (hunk 3 above)
Comment (vmsachin) [reply lib/aerospike/policy/write_policy.rb]: Have moved socket_timeout to policy

File: spec/aerospike/query_spec.rb (hunk 5 above)
Comment (vmsachin) [reply spec/aerospike/query_spec.rb]: Have changed this.

File: lib/aerospike/policy/write_policy.rb (hunk 1 above)
Comment (vmsachin) [reply lib/aerospike/policy/write_policy.rb]: I moved records_per_second to statement class as it’s the same in the Java client

File: lib/aerospike/record.rb [hunk 6]
```diff
@@ -33,6 +33,18 @@ def initialize(node, rec_key, rec_bins, rec_gen, rec_exp)
       @node = node
     end
 
+    # Arguments:
+    #   value: the key to retrieve the value for
+    #
+    # Returns:
+    #   the value of the specified key, or `nil` if `@bins` is `nil`
+    def get_value(value)
```
Comment (khaf) [inline lib/aerospike/record.rb:41]: Why is this added?

File: spec/aerospike/query_spec.rb [hunk 7]
```diff
@@ -512,6 +512,21 @@
...
+        stmt = Aerospike::Statement.new(@namespace, @set)
+        @set = "test_set_1"
+        bin = Aerospike::Bin.new("bin5", "string_value")
+        ops = [
+          Aerospike::Operation.put(bin)
+        ]
+        execute_task = client.query_execute(stmt, ops)
+        execute_task.wait_till_completed
+
+        expect(execute_task.completed?).to be true
```
Comment (khaf) [inline spec/aerospike/query_spec.rb:526]: Need to check the results to verify the changes were actually applied.

**Review by khaf — APPROVED** [review]
